`Plugin` incorrectly extends `Plugin_2` error when typechecking `vite.config.ts` with `exactOptionalPropertyTypes` flag

### Describe the bug

```
> tsc -p tsconfig.node.json

node_modules/vite/dist/node/index.d.ts:1270:26 - error TS2430: Interface 'import("C:/Users/amogus/IdeaProjects/network-chungus/node_modules/vite/dist/node/index").Plugin' incorrectly extends interface 'import("C:/Users/amogus/IdeaProjects/network-chungus/node_modules/rollup/dist/rollup").Plugin'.
  Types of property 'load' are incompatible.
    Type '((this: PluginContext, id: string, options?: { ssr?: boolean; } | undefined) => LoadResult | Promise<LoadResult>) | undefined' is not assignable to type 'LoadHook'.
      Type 'undefined' is not assignable to type 'LoadHook'.

1270 export declare interface Plugin extends Plugin_2 {
                              ~~~~~~


Found 1 error in node_modules/vite/dist/node/index.d.ts:1270
```

### Reproduction

https://stackblitz.com/edit/vitejs-vite-rtb2e8

### System Info

```shell
System:
    OS: Windows 10 10.0.19042
    CPU: (6) x64 Intel(R) Core(TM) i5-8600 CPU @ 3.10GHz
    Memory: 2.01 GB / 15.78 GB
  Binaries:
    Node: 16.9.1 - C:\Program Files\nodejs\node.EXE
    npm: 7.24.1 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Chrome: 99.0.4844.84
    Edge: Spartan (44.19041.1266.0), Chromium (99.0.1150.55)
    Internet Explorer: 11.0.19041.1566
```


### Used Package Manager

npm

### Logs

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to https://github.com/vuejs/core instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/vite/src/node/plugin.ts
+++ b/packages/vite/src/node/plugin.ts
@@ -37,7 +37,8 @@ import type { ConfigEnv, ResolvedConfig } from './'
  * If a plugin should be applied only for server or build, a function format
  * config file can be used to conditional determine the plugins to use.
  */
-export interface Plugin extends RollupPlugin {
+export interface Plugin
+  extends Omit<RollupPlugin, 'resolveId' | 'load' | 'transform'> {
   /**
    * Enforce plugin invocation tier similar to webpack loaders.
    *

--- a/packages/vite/src/node/plugin.ts
+++ b/packages/vite/src/node/plugin.ts
@@ -37,8 +37,7 @@ import type { ConfigEnv, ResolvedConfig } from './'
  * If a plugin should be applied only for server or build, a function format
  * config file can be used to conditional determine the plugins to use.
  */
-export interface Plugin
-  extends Omit<RollupPlugin, 'resolveId' | 'load' | 'transform'> {
+export interface Plugin extends RollupPlugin {
   /**
    * Enforce plugin invocation tier similar to webpack loaders.
    *
@@ -129,7 +128,7 @@ export interface Plugin
   /**
    * extend hooks with ssr flag
    */
-  resolveId?(
+  resolveId?: (
     this: PluginContext,
     source: string,
     importer: string | undefined,
@@ -140,17 +139,18 @@ export interface Plugin
        * @internal
        */
       scan?: boolean
+      isEntry: boolean
     }
-  ): Promise<ResolveIdResult> | ResolveIdResult
-  load?(
+  ) => Promise<ResolveIdResult> | ResolveIdResult
+  load?: (
     this: PluginContext,
     id: string,
     options?: { ssr?: boolean }
-  ): Promise<LoadResult> | LoadResult
-  transform?(
+  ) => Promise<LoadResult> | LoadResult
+  transform?: (
     this: TransformPluginContext,
     code: string,
     id: string,
     options?: { ssr?: boolean }
-  ): Promise<TransformResult> | TransformResult
+  ) => Promise<TransformResult> | TransformResult
 }

--- a/packages/vite/src/node/server/pluginContainer.ts
+++ b/packages/vite/src/node/server/pluginContainer.ts
@@ -97,6 +97,7 @@ export interface PluginContainer {
        * @internal
        */
       scan?: boolean
+      isEntry?: boolean
     }
   ): Promise<PartialResolvedId | null>
   transform(
@@ -527,6 +528,7 @@ export async function createPluginContainer(
       const skip = options?.skip
       const ssr = options?.ssr
       const scan = !!options?.scan
+      const isEntry = !!options?.isEntry
       const ctx = new Context()
       ctx.ssr = !!ssr
       ctx._scan = scan
@@ -546,7 +548,7 @@ export async function createPluginContainer(
           ctx as any,
           rawId,
           importer,
-          { ssr, scan }
+          { ssr, scan, isEntry }
         )
         if (!result) continue
 

--- a/packages/vite/tsconfig.check.json
+++ b/packages/vite/tsconfig.check.json
@@ -14,7 +14,9 @@
       // indirect: postcss depends on it
       "source-map-js": ["./node_modules/source-map-js/source-map.d.ts"]
     },
-    "typeRoots": []
+    "typeRoots": [],
+    "strict": true,
+    "exactOptionalPropertyTypes": true
   },
   "include": ["dist/**/*.d.ts"]
 }

